### PR TITLE
Add AMDGPU extension

### PR DESCRIPTION
**Pull request by pxl-th**
This PR adds initial support for AMDGPU via extension mechanism.

### PR Checklist

- [x] Tests are added (will add more in a bit).
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

---

- Add `FLUX_TEST_AMDGPU` env variable to specify whether to run AMDGPU tests or not.
- Add ability to switch between GPU backends using Preferences.jl.
E.g. `Flux.gpu_backend!("AMD")` or `gpu_backend="AMD"` in `LocalPreferences.toml` allows specifying which backend to use.
- Handle regular convolution, since MIOpen only supports [cross-correlation](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/convolution.html#_CPPv4N23miopenConvolutionMode_t17miopenConvolutionE) as convolution.
Therefore I plan to specify adaptor for convolution for `Flux.amd` which will flip weights during conversion once.
Not sure if this is optimal way, so if anyone has suggestions do let me know :)
- Use MIOpen for BatchNorm.

### CODE REVIEW

File: src/Flux.jl [hunk 1]
```diff
@@ -72,4 +73,22 @@ include("deprecations.jl")
...
+function gpu_backend!(backend::String)
+    backend in keys(GPU_BACKENDS) || throw(ArgumentError("""
+    Unsupported GPU backend: $backend.
+    Supported backends are: $(keys(GPU_BACKENDS)).
+    """))
+
+    @set_preferences!("gpu_backend" => backend)
+    GPU_BACKEND[] = GPU_BACKENDS[@load_preference("gpu_backend")]
+    return
+end
```
Comment (ToucheSir) [inline src/Flux.jl:92]: For type stability's sake, I think this should be done as statically as possible. The example in the Preferences.jl README looks like a good reference.

Comment (pxl-th) [reply src/Flux.jl]: Should be done.

**Comment by pxl-th** [comment]
Updated to the latest NNlib which includes AMDGPU extension as well as docs.
Should be good to go if there are no complains

File: test/amd/basic.jl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        amdgputest(bn, x; atol=1f-3, allow_nothing=true)
+    end
+end
+
+# FIXME scalar indexing. Needs NNlib.scatter?
+# @testset "Flux.onehot gpu" begin
+#     y = Flux.onehotbatch(ones(3), 1:2) |> Flux.gpu
+#     x = rand(3, 2) |> Flux.gpu
+#     @test gradient(x -> sum(x * y), x)[1] isa ROCArray
+# end
```
Comment (ToucheSir) [inline test/amd/basic.jl:94]:
```suggestion
```
The CUDA version of this test was copied over to https://github.com/FluxML/OneHotArrays.jl/blob/main/test/gpu.jl#L26 and there's a note that it doesn't seem to work on all GPUArrays. Thus we can save a few lines here.

File: ext/AMDGPUExt/conv.jl [hunk 3]
```diff
@@ -0,0 +1,9 @@
+function (c::Conv)(x::T) where T <: ROCArray
+    Flux._size_check(c, x, ndims(x) - 1 => Flux._channels_in(c))
+    σ = NNlib.fast_act(c.σ, x)
+    cdims = DenseConvDims(
+        x, c.weight; stride=c.stride, padding=c.pad,
+        dilation=c.dilation, groups=c.groups, flipkernel=true)
+    xT = Flux._match_eltype(c, x)
+    σ.(conv(xT, c.weight, cdims) .+ conv_reshape_bias(c))
+end
```
[the remark this answers is not in the document]
Comment (ToucheSir) [reply ext/AMDGPUExt/conv.jl:9]: Also, do we need to do the same for ConvTranspose?

[the remark this answers is not in the document]
Comment (pxl-th) [reply ext/AMDGPUExt/conv.jl:9]: Yes, added it.

File: test/amd/utils.jl [hunk 4]
```diff
@@ -0,0 +1,53 @@
+function amdgputest(
```
[the remark this answers is not in the document]
Comment (ToucheSir) [reply test/amd/utils.jl:1]: How about adding `allow_nothing` to normal `check_grad`? I could see it being generally useful.

File: ext/AMDGPUExt/functor.jl [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    adapt_storage(to, x), dx -> (
+        NoTangent(), NoTangent(),
+        adapt_storage(FluxAMDAdaptor(), unthunk(dx)))
+end
+
+function _amd(x)
+    check_use_amdgpu()
+    USE_AMDGPU[] ?
+        fmap(x -> Adapt.adapt(FluxAMDAdaptor(), x), x; exclude=_isleaf) :
+        x
```
Comment (ToucheSir) [inline ext/AMDGPUExt/functor.jl:37]:
```suggestion
    USE_AMDGPU[] || return x
    fmap(x -> Adapt.adapt(FluxAMDAdaptor(), x), x; exclude=_isleaf)
```

Comment (ToucheSir) [reply ext/AMDGPUExt/functor.jl:37]: Is a little cleaner.

File: ext/AMDGPUExt/functor.jl [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+adapt_storage(::FluxAMDAdaptor, x::AMDGPU.rocRAND.RNG) = x
+adapt_storage(::FluxAMDAdaptor, x::AbstractRNG) = error("""
+    Cannot map RNG of type $(typeof(x)) to AMDGPU.
+    AMDGPU execution only supports Random.default_rng().""")
+
+adapt_storage(::FluxCPUAdaptor, x::AMDGPU.rocRAND.RNG) = Random.default_rng()
+
+function ChainRulesCore.rrule(::Type{Array}, x::ROCArray)
+    Array(x), dx -> (NoTangent(), ROCArray(unthunk(dx)))
+end
```
Comment (ToucheSir) [inline ext/AMDGPUExt/functor.jl:23]:
```suggestion
```
We can remove this in line with https://github.com/FluxML/Flux.jl/pull/2167

File: test/runtests.jl [hunk 7]
```diff
@@ -60,4 +62,19 @@ Random.seed!(0)
...
+            @show AMDGPU.MIOpen.version()
+            @testset "AMDGPU" begin
+                include("amd/runtests.jl")
+            end
+        else
+            @info "AMDGPU.jl package is not functional. Skipping AMDGPU tests."
+        end
+    else
+        @info "Skipping AMDGPU tests, set FLUX_TEST_CUDA=true to run them."
+    end
```
Comment (ToucheSir) [inline test/runtests.jl:79]: Inconsistent indentation here.

File: test/runtests.jl [hunk 8]
```diff
@@ -60,4 +62,19 @@ Random.seed!(0)
...
+        if AMDGPU.functional() && AMDGPU.functional(:MIOpen)
+            @show AMDGPU.MIOpen.version()
+            @testset "AMDGPU" begin
+                include("amd/runtests.jl")
+            end
+        else
+            @info "AMDGPU.jl package is not functional. Skipping AMDGPU tests."
+        end
+    else
+        @info "Skipping AMDGPU tests, set FLUX_TEST_CUDA=true to run them."
```
Comment (ToucheSir) [inline test/runtests.jl:78]:
```suggestion
        @info "Skipping AMDGPU tests, set FLUX_TEST_AMDGPU=true to run them."
```

**Review by ToucheSir — COMMENTED** [review]
Looks real good, just a few touch-up comments from me. What are your thoughts on CI? Should we wait until the AMDGPU runners are stable before adding a Buildkite pipeline?

**Comment by pxl-th** [comment]
> Looks real good, just a few touch-up comments from me. What are your thoughts on CI? Should we wait until the AMDGPU runners are stable before adding a Buildkite pipeline?

I think we can wait until we know the CI is working and then add it here.
NNlib has configurations for AMD so we will see it there

**Review by ToucheSir — APPROVED** [review]
Ok, let's do this. Thanks @pxl-th !

**Pull request merged**